src/client/components/DicePanel.tsx

```diff
@@ -1,57 +1,382 @@
           >
-            <FontAwesomeIcon icon={icon} fixedWidth />
+            <FontAwesomeIcon icon={faPlus} fixedWidth />
           </div>
-        ))}
+        )}
       </div>
 
-      <div className={clsx("tab", active === 0 ? "active" : "")}>
-        <DiceInterface />
-        <DiceInput />
-      </div>
-      {diceCategories.map((icon, index) => (
-        <div
-          key={icon.iconName}
-          className={clsx("tab", active === index + 1 ? "active" : "")}
-        >
-          <DiceTemplates categoryIndex={index} />
-        </div>
-      ))}
+      {renderContent(active)}
     </div>
   );
 });
+
+function DiceTemplateButton({
+  category,
+  active,
+  onSetActive,
+}: {
+  category: RRDiceTemplateCategory;
+  active: boolean;
+  onSetActive: () => void;
+}) {
+  const [addMenuVisible, setAddMenuVisible] = useState(false);
+
+  return (
+    <Popover
+      interactive
+      onClickOutside={() => setAddMenuVisible(false)}
+      visible={addMenuVisible}
+      placement="bottom"
+      content={<DiceTemplateCategoryEditor category={category} />}
+    >
+      <div
+        className={clsx("tab-button", active ? "active" : "")}
+        onClick={(e) => {
+          e.stopPropagation();
+          if (e.button === 0) onSetActive();
+          else if (e.button === 2) {
+            e.stopPropagation();
+          }
+        }}
+        onContextMenu={(e) => {
+          e.stopPropagation();
+          e.preventDefault();
+          if (
+            fixedCategoryIcons.findIndex(
+              (fixedIcon) => fixedIcon === category.icon
+            ) > -1
+          )
+            return;
+          setAddMenuVisible(true);
+        }}
+        title={category.categoryName}
+      >
+        <FontAwesomeIcon icon={iconMap[category.icon]} fixedWidth />
+      </div>
+    </Popover>
+  );
+}
+
+function DiceTemplateCategoryEditor({
+  category,
+}: {
+  category: RRDiceTemplateCategory;
+}) {
+  const myself = useMyProps("diceTemplateCategories", "id");
+  const dispatch = useServerDispatch();
+
+  function deleteCategory() {
+    dispatch({
+      actions: [
+        playerDeleteDiceTemplateCategory({
+          id: myself.id,
+          categoryId: category.id,
+        }),
+      ],
+      optimisticKey: "diceTemplateCategories",
+      syncToServerThrottle: DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME,
+    });
+  }
+
+  function updateCategory(
+    category: RRDiceTemplateCategory,
+    newIcon?: typeof categoryIcons[number],
+    newName?: string
+  ) {
+    dispatch({
+      actions: [
+        playerUpdateDiceTemplateCategory({
+          id: myself.id,
+          category: {
+            id: category.id,
+            changes: {
+              icon: newIcon ?? category.icon,
+              categoryName: newName ?? category.categoryName,
+            },
+          },
+        }),
+      ],
+      optimisticKey: "diceTemplateCategories",
+      syncToServerThrottle: DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME,
```

Comment:
Should this then also be changed to 0?

Reply:
I don't think so, because it is triggered by the text input that changes the categoryName. We wouldn't want to send an update to the server on every character you type into the text input.

Reply:
Ah, fair enough